### PR TITLE
Contiguous Array Storage

### PR DESCRIPTION
**Pull request by dwerner**
This PR pulls in the `neighbors` branch which makes storage backing the HNSW contiguous, but also goes further to flatten input types to a provide a single backing `Vec<f32>` (with stride information) and therein hope to glean the benefit of contiguous array access.

I've added a `Point` implementation for `&[f32]` and optimistically I'm plumbing that through as the underlying indexing and querying.

I'm marking this as WIP for now because I want to get feedback on this before pushing it further ahead. It also still has some extra dependencies on things like `tracy` that aren't desirable to merge, but are useful for profiling. 

As far as my measurements go, the performance seems to be approximately the same as before the change. It seems that each indexing of an HNSW will produce a structure that differs in performance, and that could be due to randomized locality for graph nodes.

### CODE REVIEW

File: instant-distance/src/lib.rs [hunk 1]
```diff
@@ -383,151 +511,109 @@ where
...
+    fn new(candidate: Candidate, hnsw: &'a Hnsw) -> Self {
         Self {
             distance: candidate.distance.into_inner(),
             pid: candidate.pid,
             point: &hnsw[candidate.pid],
         }
     }
 }
-
-struct Construction<'a, P: Point> {
```
Comment (djc) [inline instant-distance/src/lib.rs:415]: Why is this change necessary?

File: instant-distance/src/lib.rs [hunk 2]
```diff
@@ -190,23 +267,66 @@ impl<'a, P, V> MapItem<'a, P, V> {
     }
 }
 
+#[derive(Default)]
+#[cfg_attr(feature = "serde", derive(Deserialize, Serialize))]
+struct PointStorage {
```
Comment (djc) [inline instant-distance/src/lib.rs:272]: Why is this abstraction needed?

If we're going to do this, this should probably be an implementation of a trait (so we can slot in an memmap-backed storage implementation later).

**Review by djc — COMMENTED** [review]
This looks like a good start and the benchmarking results sound quite encouraging!

However, in its current state this PR is very hard to review. Ideally this would:

1. Make smaller isolated commits
  * Without pulling tricks like making new modules and then merging them back in later
2. Avoid making unrelated changes (for example, the rayon imports)
3. Respect the top-down ordering
4. To the extent possible, delay the changes in the Python implementation until later commits
5. Please apply clippy suggestions (in the originating commits)

File: instant-distance-py/Cargo.toml [hunk 3]
```diff
@@ -11,12 +11,22 @@ repository = "https://github.com/InstantDomain/instant-distance"
...
+instant-distance = { version = "0.6", path = "../instant-distance", features = [
+    "with-serde",
+] }
 pyo3 = { version = "0.19.0", features = ["extension-module"] }
 serde = { version = "1", features = ["derive"] }
 serde-big-array = "0.5.0"
+
+[dev-dependencies]
+anyhow = "1.0.75"
+structopt = "0.3.24"
```
Comment (djc) [inline instant-distance-py/Cargo.toml:28]: Don't use structopt, it has been superseded by clap 3+.

File: instant-distance-py/examples/perf-fasttext.rs [hunk 4]
```diff
@@ -0,0 +1,118 @@
+//! This is a simple example of how to use the instant-distance crate to build an index from a file.
+//! The file is expected to be in the format of the fasttext word vectors.
+//!
+//! This example was built to performance using such tools as valgrind, perf, and heaptrack.
+//!
+//! Get the required file:
+//! wget https://dl.fbaipublicfiles.com/fasttext/vectors-aligned/wiki.en.align.vec
+
+use std::{
```
Comment (djc) [inline instant-distance-py/examples/perf-fasttext.rs:9]: Style nit: pelase stick to one line per module import style.

File: instant-distance-py/examples/perf-fasttext.rs [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+    #[structopt(short = "s")]
+    wait: bool,
+
+    #[structopt(short = "d", default_value = "150")]
+    width: usize,
+}
+
+fn main() -> Result<(), anyhow::Error> {
+    let opt = Opt::from_args();
+    let (words, points) = load_points(&opt.path, opt.word_count, opt.width)?;
```
Comment (djc) [inline instant-distance-py/examples/perf-fasttext.rs:47]: Please inline the single-use functions `load_points()` and `index_and_query()`.

File: instant-distance/examples/colors.rs [hunk 6]
```diff
@@ -17,6 +17,16 @@ fn main() {
 #[derive(Clone, Copy, Debug)]
 struct Point(isize, isize, isize);
 
+impl HasF32s for Point {
+    fn as_slice(&self) -> &[f32] {
+        unimplemented!("this will not work!")
```
Comment (djc) [inline instant-distance/examples/colors.rs:22]: This is a good indication that this is the wrong level of abstraction. The `Hnsw` should not care what the type of `Point` is and, by extension, should not care what the inner type used for `Point` is. Instead, we could use an associated type in the `Point` trait (maybe `type Element` or `type Inner`) instead. (`STRIDE` should probably be an associated const for `Point`.)

File: instant-distance/src/types.rs [hunk 7]
```diff
@@ -252,6 +157,202 @@ impl PointId {
     }
 }
 
+#[cfg_attr(feature = "serde", derive(Deserialize, Serialize))]
```
Comment (djc) [inline instant-distance/src/types.rs:160]: This should live near the top of the file.

**Comment by dwerner** [comment]
I've significantly reworked the original PR and API to make use of a reference type over the underlying contiguous storage. That has the SIMD code implemented for it.

I also found a curious case in the `map` unit test where some neighbors weren't being added to the search. I haven't found the cause other than I've isolated the `into_par_iter` call being implicated in `Hnsw::new`.

Performance wise, it is comparable with the SIMD code earlier. I've tried to address any requests and make this easier to review.

File: instant-distance-py/Cargo.toml [hunk 8]
```diff
@@ -11,12 +11,22 @@ repository = "https://github.com/InstantDomain/instant-distance"
 readme = "../README.md"
 
 [lib]
-name = "instant_distance"
-crate-type = ["cdylib"]
+name = "instant_distance_py"
```
Comment (djc) [inline instant-distance-py/Cargo.toml:14]: Renaming the crate (and related changes for packaging) should be in a different commit from the other changes.

File: instant-distance-py/Cargo.toml [hunk 9]
```diff
@@ -11,12 +11,22 @@ repository = "https://github.com/InstantDomain/instant-distance"
...
+crate-type = ["cdylib", "lib"]
+
+[package.metadata.maturin]
+crate-type = ["cdylib", "lib"]
 
 [dependencies]
 bincode = "1.3.1"
-instant-distance = { version = "0.6", path = "../instant-distance", features = ["with-serde"] }
+instant-distance = { version = "0.6", path = "../instant-distance", features = [
```
Comment (djc) [inline instant-distance-py/Cargo.toml:22]: Please avoid unrelated formatting changes in `Cargo.toml`.

File: instant-distance-py/examples/perf-fasttext.rs [hunk 10]
```diff
@@ -0,0 +1,111 @@
+//! This is a simple example of how to use the instant-distance crate to build an index from a file.
```
Comment (djc) [inline instant-distance-py/examples/perf-fasttext.rs:1]: Please isolate the addition of this example in a separate commit.

File: instant-distance-py/examples/perf-fasttext.rs [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+
+use clap::Parser;
+use instant_distance::{Builder, PointDataSource, PointRef, Search};
+use instant_distance_py::FloatArray;
+
+#[global_allocator]
+static ALLOC: tracy_full::alloc::GlobalAllocator = tracy_full::alloc::GlobalAllocator::new();
+
+#[derive(Parser)]
+struct Opt {
```
Comment (djc) [inline instant-distance-py/examples/perf-fasttext.rs:23]: Per the convention of top-down ordering, I'd prefer to call this `Options` and move it below `main()`. `ALLOC` should be defined at the bottom of the module.

File: instant-distance/src/contiguous.rs [hunk 12]
```diff
@@ -0,0 +1,240 @@
...
+#[cfg(feature = "serde")]
+use serde::{Deserialize, Serialize};
+
+use crate::{
+    types::{LayerId, Meta, INVALID},
+    Builder, Point, PointId,
+};
+
+pub trait PointDataSource: Sync {
+    fn decompose(&self) -> Vec<f32>;
```
Comment (djc) [inline instant-distance/src/contiguous.rs:13]: We definitely shouldn't be allocating for `decompose()`. What is the purpose of the `PointDataSource` trait and the `decompose()` method?

File: instant-distance/src/contiguous.rs [hunk 13]
```diff
@@ -0,0 +1,240 @@
...
+    fn stride() -> usize;
```
Comment (djc) [inline instant-distance/src/contiguous.rs:14]: I previously suggested you make `STRIDE` an associated const, any particular reason you didn't follow up with that?

File: instant-distance/src/contiguous.rs [hunk 14]
```diff
@@ -0,0 +1,240 @@
+use std::marker::PhantomData;
+
+use rand::{rngs::SmallRng, Rng, SeedableRng};
+#[cfg(feature = "serde")]
+use serde::{Deserialize, Serialize};
+
+use crate::{
```
Comment (djc) [inline instant-distance/src/contiguous.rs:7]: As mentioned in a previous comment, please use one import line per module.

File: instant-distance/src/lib.rs [hunk 15]
```diff
@@ -1,22 +1,26 @@
...
 use serde::{Deserialize, Serialize};
 
+mod contiguous;
+pub use contiguous::{PointDataSource, PointRef, Storage};
 mod types;
 pub use types::PointId;
-use types::{Candidate, Layer, LayerId, UpperNode, Visited, ZeroNode, INVALID};
+use types::{Candidate, Layer, LayerId, LayerSliceMut, Visited, ZeroNode};
+
+use crate::types::{Meta, INVALID};
```
Comment (djc) [inline instant-distance/src/lib.rs:23]: These should be imported from the module, no need to go through `crate::`.

File: instant-distance/src/lib.rs [hunk 16]
```diff
@@ -128,17 +136,20 @@ impl Default for Heuristic {
 }
 
 #[cfg_attr(feature = "serde", derive(Deserialize, Serialize))]
-pub struct HnswMap<P, V> {
-    hnsw: Hnsw<P>,
+pub struct HnswMap<P, V>
+where
```
Comment (djc) [inline instant-distance/src/lib.rs:140]: I don't think the bound here is necessary/helpful?

File: instant-distance/src/lib.rs [hunk 17]
```diff
@@ -172,46 +183,56 @@ where
...
 where
-    P: Point,
+    P: PointDataSource,
 {
     pub fn builder() -> Builder {
         Builder::default()
     }
 
-    fn new(points: Vec<P>, builder: Builder) -> (Self, Vec<PointId>) {
+    pub(crate) fn new(points: &[P], builder: Builder) -> (Self, Vec<PointId>) {
```
Comment (djc) [inline instant-distance/src/lib.rs:231]: Why make this `pub(crate)`?

File: instant-distance/src/lib.rs [hunk 18]
```diff
@@ -225,74 +246,35 @@ where
...
-            let next = (num as f32 * ml) as usize;
-            if next < M {
-                break;
-            }
-            sizes.push((num - next, num));
-            num = next;
-        }
-        sizes.push((num, num));
-        sizes.reverse();
-        let top = LayerId(sizes.len() - 1);
```
Comment (djc) [inline instant-distance/src/lib.rs:250]: It would probably be nice to have a separate commit that moves this into the `types` module, so that it can later morph into the `Meta` type (and in that transition make it easier to review what changed to get there).

File: instant-distance/src/lib.rs [hunk 19]
```diff
@@ -301,78 +283,83 @@ where
...
-                range.into_iter().for_each(|i| inserter(PointId(i as u32)))
-            } else {
-                range
-                    .into_par_iter()
-                    .for_each(|i| inserter(PointId(i as u32)));
-            }
+            let Range { start, end } = state.meta.points(layer);
+
+            // FIXME:
+            // There is some sort of race here where neighbors is missing nodes.
```
Comment (djc) [inline instant-distance/src/lib.rs:295]: We won't be able to merge this. Can you debug the issue?

File: instant-distance/src/lib.rs [hunk 20]
```diff
@@ -425,7 +418,10 @@ struct Construction<'a, P: Point> {
     done: AtomicUsize,
 }
 
-impl<'a, P: Point> Construction<'a, P> {
+impl<'a, P> Construction<'a, P>
```
Comment (djc) [inline instant-distance/src/lib.rs:421]: Style nit: I prefer declaring with inline bounds where possible, as before.

File: instant-distance/src/contiguous.rs [hunk 21]
```diff
@@ -0,0 +1,240 @@
...
+    pub fn from_data(values: &'a [f32]) -> Self {
+        Self(values)
+    }
+}
+
+#[cfg_attr(feature = "serde", derive(Deserialize, Serialize))]
+#[derive(Default)]
+pub struct ContiguousStorage<T: PointDataSource> {
+    pub values: Vec<f32>,
+    pub order: Vec<usize>,
```
Comment (djc) [inline instant-distance/src/contiguous.rs:57]: What is `order` for?

File: instant-distance/examples/colors.rs (hunk 6 above)
Comment (dwerner) [reply instant-distance/examples/colors.rs]: I did try a trait with an associated type `Element` (which had an associated const for STRIDE as well) for a while but that made resolving down to an f32 (or whatever type `Element` was) much harder to work with trait bounds for the SIMD implementation. I instead opted to, in the current iteration, reduce this down to just f32, since our internal types `Item, MapItem` were effectively limiting us to that anyway (due to our use of `OrderedFloat`). If you see a way where I could have a purely generic `Element` here, I'd prefer that as well.

File: instant-distance-py/Cargo.toml (hunk 9 above)
Comment (dwerner) [reply instant-distance-py/Cargo.toml:22]: vscode needs reigned in, as in imports as well.

File: instant-distance/src/contiguous.rs (hunk 12 above)
Comment (dwerner) [reply instant-distance/src/contiguous.rs:13]: I agree this would be better as a slice, but it's worth touching here to get a better idea for the intended design. 

`PointDataSource` is effectively the input type to our `Hnsw`, and is intended to be a way to convert what it's `impl`'d on to something that we can store, and later wrap in our "output" (or `Point` equivalent) type. In this iteration, that's `PointRef`. This means that the point data needs to be convertible (or "decomposed") to our internal storage (which for now is `Vec<f32>`).

Some open questions here:

- Prospectively, what does an API here look like that provides a generic input type that we can consume, and also provide access to, perhaps as a wrapper over slices of `f32` (or an `Element`, if that is workable). I'm also trying to keep the SIMD code workable.
- With the `Point` trait, we allowed `impl` of `fn distance` itself, but if we are looking to provide a general implementation internally (?), is that a concern that should remain externally-implementable? I could be misunderstanding the general goal for the API in this regard.

File: instant-distance/src/contiguous.rs (hunk 13 above)
Comment (dwerner) [reply instant-distance/src/contiguous.rs:14]: In actuality I did have an associated const at one point. I migrated to a static function  during an exploratory change, but I'm happy to move back to a const.

File: instant-distance/src/lib.rs (hunk 16 above)
Comment (dwerner) [reply instant-distance/src/lib.rs:140]: Leftovers from more exploratory changes, missed it.

File: instant-distance/src/lib.rs (hunk 17 above)
Comment (dwerner) [reply instant-distance/src/lib.rs:231]: No reason anymore.

File: instant-distance/src/lib.rs (hunk 18 above)
Comment (dwerner) [reply instant-distance/src/lib.rs:250]: Ok I'll take this into account when I do another rebase.

File: instant-distance/src/lib.rs (hunk 19 above)
Comment (dwerner) [reply instant-distance/src/lib.rs:295]: I plan to spend some more time on this.

File: instant-distance/src/contiguous.rs (hunk 21 above)
Comment (dwerner) [reply instant-distance/src/contiguous.rs:57]: In the original code from the `neighbors` branch, nodes are reordered and assigned layers. This is then stored. I wanted to preserve the reordering, but not mutate the underlying store, which I believe will be needed when we get to mmap-ing or working with point data from something like lmdb. This introduces an indirection, but in my testing, hasn't negatively impacted performance. This is another design consideration I'm happy to hear your thoughts on.

**Comment by dwerner** [comment]
> Definitely a step in the right direction. I think there is still a bunch of work to do in making the design more robust and making the changes easier to review -- as well as potentially digging out the source of the test failures you found.

Thanks for the review! I've got a few questions around design goals. I appreciate the input.

File: instant-distance/examples/colors.rs (hunk 6 above)
Comment (djc) [reply instant-distance/examples/colors.rs]: `Item` and `MapItem` apply `OrderedFloat` to the `distance` -- it's fine for `Point::distance()` to still yield `f32`, but I don't think that needs to constrain the data internal to the points?

File: instant-distance/src/contiguous.rs (hunk 12 above)
Comment (djc) [reply instant-distance/src/contiguous.rs:13]: At the Rust level, I think we can make `Point` pretty abstract: it provides a `distance() -> f32` method, an associated `Element` type and a `STRIDE` const. As I understand it, this should be sufficient on the Rust side. I suppose we can probably dump the `Clone` bound and do something where either `Point` exposes `as_slice(&self) -> &[Self::Element]` which we can then use to write in the `Hnsw::points` (or equivalent storage) or we can do it the other way around and add a `Point` method `store(&self, data: &mut [Self::Element])` method.

File: instant-distance/src/contiguous.rs (hunk 21 above)
Comment (djc) [reply instant-distance/src/contiguous.rs:57]: I think we have to randomize the order points on construction, and then make sure to avoid further reordering after construction. The way ordering works now is that we give the caller a `Vec<PointId>` after construction which avoids the `Hnsw` having to keep track of the original order, IIRC.

File: instant-distance/src/lib.rs [hunk 22]
```diff
@@ -225,74 +246,35 @@ where
...
-        sizes.reverse();
-        let top = LayerId(sizes.len() - 1);
+        let mut meta = Meta::new(ml, points.len());
 
         // Give all points a random layer and sort the list of nodes by descending order for
         // construction. This allows us to copy higher layers to lower layers as construction
         // progresses, while preserving randomness in each point's layer and insertion order.
 
-        assert!(points.len() < u32::MAX as usize);
-        let mut shuffled = (0..points.len())
```
[the remark this answers is not in the document]
Comment (dwerner) [reply instant-distance/src/lib.rs:257]: Done

File: instant-distance/src/contiguous.rs (hunk 14 above)
Comment (dwerner) [reply instant-distance/src/contiguous.rs]: Fixed in a later commit

File: instant-distance/src/types.rs (hunk 7 above)
Comment (dwerner) [reply instant-distance/src/types.rs:14]: Done

File: instant-distance-py/Cargo.toml (hunk 8 above)
Comment (dwerner) [reply instant-distance-py/Cargo.toml:14]: Should be in a separate commit now.

File: instant-distance-py/examples/perf-fasttext.rs (hunk 10 above)
Comment (dwerner) [reply instant-distance-py/examples/perf-fasttext.rs:1]: Done

File: instant-distance-py/examples/perf-fasttext.rs (hunk 11 above)
Comment (dwerner) [reply instant-distance-py/examples/perf-fasttext.rs]: Done

File: instant-distance/src/contiguous.rs (hunk 12 above)
Comment (dwerner) [reply instant-distance/src/contiguous.rs]: Thanks, that was helpful.

File: instant-distance/src/lib.rs (hunk 18 above)
Comment (dwerner) [reply instant-distance/src/lib.rs:250]: Done.

File: instant-distance/src/lib.rs (hunk 19 above)
Comment (dwerner) [reply instant-distance/src/lib.rs]: I resolved this (in [74d697c](https://github.com/instant-labs/instant-distance/pull/47/commits/74d697ce078f9c2f577e14a0d5f70ff80046fc6d))

File: instant-distance/src/lib.rs (hunk 20 above)
Comment (dwerner) [reply instant-distance/src/lib.rs]: Fixed.

File: instant-distance/src/contiguous.rs (hunk 21 above)
Comment (dwerner) [reply instant-distance/src/contiguous.rs]: Ok `order` is gone now and we keep the backing store shuffled as before.

**Comment by dwerner** [comment]
I spent some time today and tried to make this easier to review.
